### ISSUE TITLE
Unable to provide EngineAuthData from /etc/ecs_container_agent/config.json

### ISSUE THREAD

**Issue by ejholmes** [aws/amazon-ecs-agent#732]
When trying to move from environment variables to a config.json file, I noticed that the agent kept coming up with a bad docker auth configuration (logging `[WARN] Could not parse 'dockercfg' type auth config`) regardless of having a properly formatted config.json, like this:

```json
{
  "Cluster": "<redacted>",
  "EngineAuthType": "dockercfg",
  "EngineAuthData": {
    "https://index.docker.io/v1/":{
      "auth":"<redacted>",
      "email":"<redacted>"
    }
  },
  "Checkpoint": true,
  "DataDir": "/var/run/ecs_agent",
  "TaskIAMRoleEnabled": true,
  "InstanceAttributes": {
    "hostgroup": "<redacted>"
  }
}
```

I tried a little Go program to figure out what the agent was using for the `EngineAuthData`:

```go
package main

import (
	"fmt"
	"log"

	"github.com/aws/amazon-ecs-agent/agent/config"
	"github.com/aws/amazon-ecs-agent/agent/ec2"
	"github.com/aws/amazon-ecs-agent/agent/engine/dockerauth"
)

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

func run() error {
	ec2MetadataClient := ec2.NewBlackholeEC2MetadataClient()
	cfg, err := config.NewConfig(ec2MetadataClient)
	if err != nil {
		return err
	}
	fmt.Printf("%#v\n", cfg)
	fmt.Println(string(cfg.EngineAuthData.Contents()))
	auth := dockerauth.NewDockerAuthProvider(cfg.EngineAuthType, cfg.EngineAuthData.Contents())
	fmt.Println(auth)
	return nil
}
```

Which was printing an empty string for `cfg.EngineAuthData.Contents()`. I believe this happens because the Config object returned from pulling values out of environment variables always sets a non-nil value: https://github.com/aws/amazon-ecs-agent/blob/v1.14.0/agent/config/config.go#L333.

A simple fix for this particular instance would be to set `EngineAuthData` to nil of `ECS_ENGINE_AUTH_DATA` is an empty string, which would then fallback to the value provided by the config. However, as a user, I'd prefer to be able to pass a path to a config file, and not have the agent try to merge in environment variables.

**Comment by richardpen** [comment]
@ejholmes ECS Agent accept two types of docker auth data: `docker` and `dockercfg`. The `EngineAuthType: docker` accepts the format of json, and `EngineAuthType: dockercfg` accepts the format of string, see document [here](https://godoc.org/github.com/aws/amazon-ecs-agent/agent/engine/dockerauth). In your configuration you want to use the auth data generated by `docker login` then you should pass the parameter as string instead of json, something like this should work:
```
{
  "Cluster": "<redacted>",
  "EngineAuthType": "dockercfg",
  "EngineAuthData": '{"https://index.docker.io/v1/":{"auth":"<redacted>","email":"<redacted>"}',
  },
  "Checkpoint": true,
  "DataDir": "/var/run/ecs_agent",
  "TaskIAMRoleEnabled": true,
  "InstanceAttributes": {
    "hostgroup": "<redacted>"
  }
}
```

Please let me know if you still experience the problem after this change.

Thanks,
Richard

**Comment by richardpen** [comment]
@ejholmes Thanks for reporting this issue, after some investigation we found this is an issue in the agent code. We will work on the fix and keep you updated.

Thanks,
Richard

**Comment by samuelkarp** [comment]
Released in [v1.14.2](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.14.2).
